flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepointTest.java

```diff
@@ -0,0 +1,248 @@
+                new ExecutionGraphHandler(
+                        executionGraph,
+                        log,
+                        ctx.getMainThreadExecutor(),
+                        ctx.getMainThreadExecutor());
+        OperatorCoordinatorHandler operatorCoordinatorHandler =
+                new OperatorCoordinatorHandler(
+                        executionGraph,
+                        (throwable) -> {
+                            throw new RuntimeException("Error in test", throwable);
+                        });
+
+        CheckpointCoordinator checkpointCoordinator =
+                new MockCheckpointCoordinator(savepointFuture);
+
+        StopWithSavepoint stopWithSavepoint =
+                new StopWithSavepoint(
+                        ctx,
+                        executionGraph,
+                        executionGraphHandler,
+                        operatorCoordinatorHandler,
+                        log,
+                        ClassLoader.getSystemClassLoader(),
+                        checkpointCoordinator,
+                        "savepointTargetDir",
+                        false,
+                        operationCompletionFuture);
+        return new StopWithSavepointEnvironment(
+                stopWithSavepoint,
+                executionGraphTerminationFuture,
+                operationCompletionFuture,
+                savepointFuture);
+    }
+
+    private static class MockCheckpointCoordinator extends CheckpointCoordinator {
+        private static ExecutionVertex vertex = mockExecutionVertex(new ExecutionAttemptID());
+        private static ExecutionVertex[] defaultVertices = new ExecutionVertex[] {vertex};
+        private final CompletableFuture<CompletedCheckpoint> savepointFuture;
+
+        MockCheckpointCoordinator(CompletableFuture<CompletedCheckpoint> savepointFuture) {
+            super(
+                    new JobID(),
+                    new CheckpointCoordinatorConfiguration
+                                    .CheckpointCoordinatorConfigurationBuilder()
+                            .build(),
+                    Collections.emptyList(),
+                    new StandaloneCheckpointIDCounter(),
+                    new StandaloneCompletedCheckpointStore(1),
+                    new MemoryStateBackend(),
+                    Executors.directExecutor(),
+                    new CheckpointsCleaner(),
+                    new ManuallyTriggeredScheduledExecutor(),
+                    SharedStateRegistry.DEFAULT_FACTORY,
+                    new CheckpointFailureManager(0, NoOpFailJobCall.INSTANCE),
+                    new CheckpointPlanCalculator(
+                            new JobID(),
+                            Arrays.asList(defaultVertices),
+                            Arrays.asList(defaultVertices),
+                            Arrays.asList(defaultVertices)),
+                    new ExecutionAttemptMappingProvider(Arrays.asList(defaultVertices)));
```

Comment:
The pain to set these things up should be a good indicator that we are doing something wrong here.